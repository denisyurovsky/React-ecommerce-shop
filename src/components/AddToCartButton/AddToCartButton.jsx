import { Button } from '@mui/material';
import PropTypes from 'prop-types';
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { toast } from 'react-toastify';

import { notificationError } from '../../helpers/constants/constants';
import { findProductIndexById } from '../../helpers/utils/findProductIndexById';
import { pageView } from '../../pages/ProductListPage/constants/constants';
import {
  selectCart,
  addProduct,
  decreaseProduct,
} from '../../store/cart/cartSlice';

import listStyles from './AddToCartButtonList.module.scss';
import moduleStyles from './AddToCartButtonModule.module.scss';

export const AddToCartButton = ({
  product,
  viewMode = pageView.MODULE_VIEW,
}) => {
  const cart = useSelector(selectCart);
  const products = cart.sellers[product.userId]
    ? cart.sellers[product.userId].products
    : [];

  const dispatch = useDispatch();
  const index = findProductIndexById(products, product.id);

  const addHandler = () => {
    dispatch(addProduct({ product })).then(() => {
      if (cart.errorOccurred) {
        toast.error(notificationError);
      }
    });
  };

  const decreaseHandler = () => {
    dispatch(decreaseProduct({ product })).then(() => {
      if (cart.errorOccurred) {
        toast.error(notificationError);
      }
    });
  };

  const styles = viewMode === pageView.MODULE_VIEW ? moduleStyles : listStyles;

  return products[index] ? (
    <div className={styles.container}>
      <Button
        variant="contained"
        className={styles.button + ' ' + styles.leftButton}
        onClick={decreaseHandler}
      >
        -
      </Button>
      <p className={styles.counter}>{products[index].quantity}</p>
      <Button
        variant="contained"
        className={styles.button + ' ' + styles.rightButton}
        onClick={addHandler}
      >
        +
      </Button>
    </div>
  ) : (
    <div className={styles.bigButtonContainer}>
      <Button
        variant="contained"
        className={styles.bigButton}
        onClick={addHandler}
      >
        + add to cart
      </Button>
    </div>
  );
};

AddToCartButton.propTypes = {
  product: PropTypes.shape({
    userId: PropTypes.number,
    quantity: PropTypes.number,
    createdAt: PropTypes.string,
    updatedAt: PropTypes.string,
    name: PropTypes.string,
    discountPrice: PropTypes.number,
    price: PropTypes.number,
    id: PropTypes.number,
    category: PropTypes.shape({
      name: PropTypes.string,
    }),
  }).isRequired,
  viewMode: PropTypes.oneOf([pageView.LIST_VIEW, pageView.MODULE_VIEW]),
};
