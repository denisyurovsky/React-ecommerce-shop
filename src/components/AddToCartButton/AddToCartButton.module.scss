@import "../../assets/styles/colors.scss";

.container {
  min-width: 120px;
  width: 150px;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
}

.rightButton {
  border-radius: 0 5px 5px 0 !important;
}

.leftButton {
  line-height: 21px !important;
  font-size: 27px !important;
  border-radius: 5px 0 0 5px !important;
}

.button {
  height: 37px !important;
  width: 54px !important;
  min-width: unset !important;
  background-color: $primaryMain;
  text-transform: capitalize;
}

.bigButton {
  width: 150px;
  background-color: $primaryMain;
  text-transform: capitalize;
}

.counter {
  margin: 0;
  height: 37px;
  padding-top: 9px;
  font-size: 14px;
  text-align: center;
  vertical-align: center;
  width: 100%;
  border-bottom: 1px solid $greyDark;
  border-top: 1px solid $greyDark;
}

@media all and (max-width: 400px) {
  .bigButton {
    font-size: 10px !important;
    width: 120px;
  }

  .container {
    width: 120px;
  }
}
