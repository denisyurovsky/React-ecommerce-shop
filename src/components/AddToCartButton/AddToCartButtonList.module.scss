@import "../../assets/styles/colors.scss";
@import "../../assets/styles/breakpoints.scss";

.container {
  min-width: 120px;
  width: 160px;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  height: 36px;
}

.container,
.bigButtonContainer {
  padding-right: 1rem;
}

.rightButton {
  border-radius: 0 5px 5px 0 !important;
}

.leftButton {
  line-height: 21px !important;
  font-size: 27px !important;
  border-radius: 5px 0 0 5px !important;
}

.button {
  width: 60px !important;
  min-width: unset !important;
  background-color: $primaryMain;
  text-transform: capitalize !important;
}

.bigButton {
  width: 150px;
}

.counter {
  margin: 0;
  padding-top: 8px;
  font-size: 14px;
  text-align: center;
  vertical-align: center;
  width: 100%;
  border-bottom: 1px solid $greyDark;
  border-top: 1px solid $greyDark;
}

.counter,
.button,
.bigButton {
  height: 36px;
}
