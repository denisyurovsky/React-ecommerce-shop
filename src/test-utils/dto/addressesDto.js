export const newAddress = {
  title: 'Mr.',
  name: 'Jon',
  surname: 'Smith',
  country: { id: 'RU', name: 'Russian Federation' },
  city: 'Saint Petersburg',
  street: 'Nevskiy',
  building: '8',
  flat: '33',
  phone: '+7 (981) 1112233',
  zip: '168142',
};

export const addresses = [
  {
    id: 1,
    title: 'Mr.',
    name: 'Alexander',
    surname: 'Sidorov',
    country: {
      id: 'RU',
      name: 'Russian Federation',
    },
    city: 'Saint Petersburg',
    street: 'Pulkovskaya',
    building: '8',
    flat: '4',
    phone: '+7 (981) 3002020',
    zip: '168142',
  },
  {
    id: 2,
    title: 'Mr.',
    name: 'Ivan',
    surname: 'Ivanov',
    country: {
      id: 'RU',
      name: 'Russian Federation',
    },
    city: 'Saint Petersburg',
    street: 'Leninskiy',
    building: '8',
    flat: '4',
    phone: '+7 (981) 1112233',
    zip: '168142',
  },
  {
    id: 3,
    title: 'Mrs.',
    name: 'Olga',
    surname: 'Petrova',
    country: {
      id: 'RU',
      name: 'Russian Federation',
    },
    city: 'Saint Petersburg',
    street: 'Pobedy',
    building: '8',
    flat: '4',
    phone: '+7 (981) 1233232',
    zip: '168142',
  },
  {
    id: 4,
    ...newAddress,
  },
];
