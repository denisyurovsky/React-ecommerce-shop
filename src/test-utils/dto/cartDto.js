export const testCart = {
  products: [
    { quantity: 1, checked: true, productId: 1, price: 790 },
    { quantity: 2, checked: true, productId: 2, price: 638 },
    { quantity: 1, checked: true, productId: 3, price: 67 },
  ],
  isLoading: false,
  errorOccurred: false,
  errorMessage: '',
  totalQuantity: 4,
  totalPrice: 1505,
};
