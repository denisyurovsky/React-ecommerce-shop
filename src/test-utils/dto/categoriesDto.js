const categoriesDto = [
  'All categories',
  'Grocery',
  'Industrial',
  'Clothing',
  'Computers',
];

export default categoriesDto;
