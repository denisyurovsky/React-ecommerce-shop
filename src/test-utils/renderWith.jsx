import { configureStore } from '@reduxjs/toolkit';
import { render as rtlRender } from '@testing-library/react';
import PropTypes from 'prop-types';
import React from 'react';
import { Provider } from 'react-redux';
import { ToastContainer } from 'react-toastify';

import cartReducer from '../store/cart/cartSlice';
import categoriesReducer from '../store/categories/categoriesSlice';
import feedbackReducer from '../store/feedback/feedbackSlice';
import productsReducer from '../store/products/productsSlice';
import userReducer from '../store/user/userSlice';

import RouterConnected from './RouterConnected';

function render(
  component,
  preloadedState = {},
  path = '/',
  initialPaths = ['/']
) {
  const store = configureStore({
    reducer: {
      cart: cartReducer,
      products: productsReducer,
      categories: categoriesReducer,
      user: userReducer,
      feedback: feedbackReducer,
    },
    preloadedState,
  });

  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <RouterConnected
          component={children}
          path={path}
          initialPaths={initialPaths}
        />
        <ToastContainer />
      </Provider>
    );
  }

  Wrapper.propTypes = {
    children: PropTypes.object,
  };

  return rtlRender(component, { wrapper: Wrapper });
}

export * from '@testing-library/react';
export default render;
