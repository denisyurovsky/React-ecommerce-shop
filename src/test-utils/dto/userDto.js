import { cartItems } from './cartDto';

export const userDto = {
  email: 'IvanIvanov@yandex.ru',
  firstName: 'Ivan',
  lastName: 'Ivanov',
  cart: cartItems,
  id: 1,
};

export const userWithEmptyCart = {
  name: 'name',
  id: 1,
  email: 'email@email.com',
  cart: {
    sellers: {},
    totalQuantity: 0,
    totalPrice: 0,
  },
};
