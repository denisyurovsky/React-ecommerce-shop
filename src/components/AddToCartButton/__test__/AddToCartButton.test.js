import { fireEvent, render, screen } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import React from 'react';

import { testCart } from '../../../test-utils/dto/cartDto';
import testCards from '../../../test-utils/dto/productsDto';
import { userDto } from '../../../test-utils/dto/userDto';
import renderWith from '../../../test-utils/renderWith';
import { AddToCartButton } from '../AddToCartButton';

const serverUser = {
  ...userDto,
};

const initialUser = {
  user: {
    id: 1,
  },
};

const preloadedState = {
  user: initialUser,
  cart: testCart,
};

const successfulHandlers = [
  rest.put('/cart/1', (req, res, ctx) => {
    serverUser.cart = req.body.cart;

    return res(ctx.status(200), ctx.json(serverUser));
  }),
  rest.get('/users/1', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(serverUser));
  }),
];

describe('AddToCartButton component', () => {
  describe('snapshots', () => {
    it('renders a valid snapshot', () => {
      const { asFragment } = renderWith(
        <AddToCartButton product={testCards[0]} />,
        preloadedState
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('valid behaviour of buttons when server responds', () => {
    const server = setupServer(...successfulHandlers);

    beforeAll(() => {
      server.listen();
    });
    afterAll(() => server.close());
    it('reacts on + clicks correctly', async () => {
      renderWith(<AddToCartButton product={testCards[2]} />, preloadedState);

      const addButton = await screen.findByText('+ add to cart');

      fireEvent.click(addButton);
      const valueFromServer = await screen.findByText('1');

      expect(valueFromServer).toBeInTheDocument();
    });

    it('reacts on - clicks correctly', async () => {
      renderWith(<AddToCartButton product={testCards[2]} />, preloadedState);

      const addButton = await screen.findByText('+ add to cart');

      fireEvent.click(addButton);

      const secondAddButton = await screen.findByText('+');

      fireEvent.click(secondAddButton);

      expect(await screen.findByText('2')).toBeInTheDocument();
      const decreaseButton = await screen.findByText('-');

      fireEvent.click(decreaseButton);

      expect(await screen.findByText('1')).toBeInTheDocument();
    });
  });
});
