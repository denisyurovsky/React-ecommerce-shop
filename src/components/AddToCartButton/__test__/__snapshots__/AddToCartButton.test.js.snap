// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddToCartButton component snapshots renders a valid snapshot 1`] = `
<DocumentFragment>
  <div
    class="bigButtonContainer"
  >
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root bigButton css-sghohy-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      + add to cart
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    class="Toastify"
  />
</DocumentFragment>
`;
